FeedStoreChallenge/CoreData/CoreDataFeedStore.swift

```
@@ -0,0 +1,115 @@
+		}
+		persistentContainer = NSPersistentContainer(dataModelName: dataModelName, model: model, storeURL: storeURL)
+		managedContext = persistentContainer.newBackgroundContext()
+	}
+	
+	// MARK: - FeedStore
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		let context = self.managedContext
+		
+		context.perform {
+			do {
+				try self.deleteCache()
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		let context = self.managedContext
+		
+		context.perform {
+			do {
+				try self.deleteCache()
+				let _ = CoreDataFeedMapper.mapToStorableFeed(feed: feed, timestamp: timestamp, in: context)
+				try context.save()
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		let context = self.managedContext
+		
+		context.perform {
+			do {
+				guard let cachedFeed = try context.fetch(CDFeed.fetchRequest()).first as? CDFeed else {
+					completion(.empty)
+					return
+				}
+				let (feed, timestamp) = CoreDataFeedMapper.mapToFeed(cachedFeed)
+				completion(.found(feed: feed, timestamp: timestamp))
+			} catch {
+				completion(.failure(error))
+			}
+		}
+	}
+	
+	// MARK: - Helpers
+	
+	private func deleteCache() throws {
+		if let feedCache = try managedContext.fetch(CDFeed.fetchRequest()).first as? CDFeed {
+			managedContext.delete(feedCache)
+			try managedContext.save()
+		}
+	}
+}
+
+private extension NSPersistentContainer {
+	convenience init(dataModelName: String, model: NSManagedObjectModel, storeURL: URL) {
+		let description = NSPersistentStoreDescription(url: storeURL)
+		self.init(name: dataModelName, managedObjectModel: model)
+		persistentStoreDescriptions = [description]
+		loadPersistentStores { _, _ in }
```

`loadPersistentStores` can fail, but the result is not handled.

This means this class may be instantiated in an invalid state - leading to undefined behavior at runtime.

Ideally, if it fails to `loadPersistentStores`, it should throw an error so clients can handle it.

---

I've redone this part now, after going through the last lecture of the Persistence chapter (very nice).
Decided to replace the `NSPersistentContainer` nullable initializer with a static func `load` that throws, that way we can better handle the error cases (didn't want to have an init that can return nil and throw, that is just confusing).